stdlib/Markdown/src/render/terminal/render.jl

```diff
@@ -16,118 +16,145 @@ end
+        bullet = isordered(md) ? "$(i + md.ordered - 1)." : "• "
+        print(io, ' '^2margin, styled"{markdown_list:$bullet} ")
+        content = annotprint(term, point, columns - 10)
+        lines = split(rstrip(content), '\n')
+        for (l, line) in enumerate(lines)
+            l > 1 && print(io, ' '^(2margin+3))
+            print(io, lstrip(line))
+            l < length(lines) && println(io)
         end
-        i < lastindex(md.items) && print(io, '\n', '\n')
-    end
-end
-
-function _term_header(io::IO, md, char, columns)
-    text = terminline_string(io, md.text)
-    with_output_color(:bold, io) do io
-        pre = ' '^margin
-        print(io, pre)
-        line_no, lastline_width = print_wrapped(io, text,
-                                                width=columns - 4margin; pre)
-        line_width = min(lastline_width, columns)
-        if line_no > 1
-            line_width = max(line_width, div(columns, 3)+length(pre))
-        end
-        header_width = max(0, line_width-length(pre))
-        char != ' ' && header_width > 0 && print(io, '\n', ' '^(margin), char^header_width)
+        i < length(md.items) && print(io, '\n'^(1 + md.loose))
     end
 end
 
 const _header_underlines = collect("≡=–-⋅ ")
 # TODO settle on another option with unicode e.g. "≡=≃–∼⋅" ?
 
-function term(io::IO, md::Header{l}, columns) where l
+function term(io::AnnotIO, md::Header{l}, columns) where l
+    face = Symbol("markdown_h$l")
     underline = _header_underlines[l]
-    _term_header(io, md, underline, columns)
+    pre = ' '^margin
+    local line_width
+    with_output_annotations(io, :face => face) do io
+        headline = annotprint(terminline, md.text)
+        lines = wraplines(headline, columns - 4margin)
+        for (i, line) in enumerate(lines)
+            print(io, pre, line)
+            i < length(lines) && println(io)
+        end
+        line_width = min(textwidth(lines[end]), columns)
+        if length(lines) > 1
+            line_width = max(textwidth(lines[end]), div(columns, 3)+length(pre))
+        end
+    end
+    header_width = max(0, line_width)
+    if underline != ' ' && header_width > 0
+        print(io, '\n', ' '^(margin))
+        with_output_annotations(io -> print(io, underline^header_width), io, :face => face)
+    end
 end
 
 function term(io::IO, md::Code, columns)
-    with_output_color(:cyan, io) do io
-        L = lines(md.code)
-        for i in eachindex(L)
-            print(io, ' '^margin, L[i])
-            i < lastindex(L) && println(io)
+    code = if md.language ∈ ("", "julia")
+        highlight(md.code)
+    elseif md.language == "julia-repl" || Base.startswith(md.language, "jldoctest")
+        hl = AnnotatedString(md.code)
+        for (; match) in eachmatch(r"(?:^|\n)julia>", hl)
+            StyledStrings.face!(match, :repl_prompt_julia)
+            afterprompt = match.offset + match.ncodeunits + 1
+            _, exprend = Meta.parse(md.code, afterprompt, raise = false)
```

Comment:
It feels a bit tricky to have a full REPL here, but anyways, I think this is supposed to use `Base.parse_input_line` aka `Meta.parseall` here and repeatedly add lines until the expression is complete. The result may not be quite the same as `eachmatch` would return, but may be closer to what `eachline` returns, followed by `if ismatch` followed by `Meta.parseall` consuming more lines until it reaches a concluding statement (a blank line or a complete expression or a parse error)

Reply:
This is modelled after the current behaviour of `OhMyREPL` which seems to work. I think Kristoffer and I had a brief conversation on this on Slack where I asked "why `Meta.parse` not `Meta.parseatom`?" but I forget the details and it's been swallowed by the Slack-hole at this point :smiling_face_with_tear:.

I'll also note that it's actually somewhat difficult to tell the behavioural difference between `parse`, `parseatom`, and `parseall` as the latter two have no docstrings and just call `Core._parse`, which itself seems to have no (useful) comments.